Modified initialization of SimpleCar And Default Number of TrajectoryCars in AutomotiveDemo.

liangfok opened the pull request:
Prior to this PR, there would always be at least one `SimpleCar` in the demo, which is at times undesirable when the focus is on other types of vehicles. In addition, when multiple `SimpleCar` instances are added, they would all initially be at exactly the same location.

This PR removes the one-`SimpleCar` minimum and initializes the `SimpleCar` with a y-offset so they're not initially at the same location.

This PR also changes the default number of `TrajectoryCar` instances to be zero instead of one. This is so we don't have to include a `--num_trajectory_car=0` command line flag when we're testing other vehicles.

Here's the initial state of the simulation when two `SimpleCar` instances are added:

```
$ bazel run drake/automotive:demo -- --simple_car_names=alice,bob \
    --driving_command_gui_names=alice,bob --num_trajectory_car=1
```

<img width="893" alt="screen shot 2017-04-08 at 11 57 54 pm" src="https://cloud.githubusercontent.com/assets/1388098/24834601/b9d2dab2-1cb8-11e7-8a82-db676b0bb022.png">


# Video

https://youtu.be/n0MZcgPLvVM

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5774)
<!-- Reviewable:end -->

liangfok commented:
+@jwnimmer-tri for all review please. Thanks.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-:-KhCZeWkXUWXGTQsJ3jg:b-3sfv6e)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhDe70XPVTHDrqEhDfN:-KhDe70Z4dugfCBFsi7O:b-f4wrf6) (no related file):*
I am surprised not to see a `README.md` update.   Surely this breaks the first demo (because we say it will show one ego car), and the UI control driver (due to the new lcm_tag that no longer matches the simulation)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I'm increasing the priority because this PR contains an actual bug fix. It fixes a bug in the steering command driver that prevents it from steering the vehicle (the steering angle was always 0).

liangfok commented:
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhDe70XPVTHDrqEhDfN:-KhFu-JFfHA6Dw4TypYO:b-qsvnoa) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I am surprised not to see a `README.md` update.   Surely this breaks the first demo (because we say it will show one ego car), and the UI control driver (due to the new lcm_tag that no longer matches the simulation)?
</blockquote></details>

Done. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@rpoyner-tri for feature review since this now modifies `automotive_demo.py`.

---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-:-KhFu1jv5DLm8UFm9IEn:b7cr5mg)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/automotive_demo.py, line 213 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMx59o5viKuGPL3Qfm:-KhMx59o5viKuGPL3Qfn:bw3qn06) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/automotive_demo.py#L213)):*
> ```Python
>                         default=False,
>                         help="print commands instead of running them")
>     parser.add_argument("--simple_car_names", default="")
> ```

Missing docs, etc., etc.

---

*[drake/automotive/automotive_demo.py, line 240 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMx9oQ1qwGbwhTlAWO:-KhMx9oQ1qwGbwhTlAWP:bwrq1kz) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/automotive_demo.py#L240)):*
> ```Python
>                 wait_for_lcm_message_on_channel('DRAKE_VIEWER_STATUS')
> 
>         if args.simple_car_names != "":
> ```

FYI I think just `if args.simple_car_names:` would be more pythonic.

---

*[drake/automotive/automotive_demo.py, line 245 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMwXLcsxRHI51wkS0Q:-KhMwXLcsxRHI51wkS0R:b-st4eyr) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/automotive_demo.py#L245)):*
> ```Python
> 
>         if args.launch_steering_command_driver:
>             if args.simple_car_names != "":
> ```

FYI It would be more pythonic to rollup this condition with the line above with an `and`.

---

*[drake/automotive/README.md, line 45 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMwHRUn5Yz9mUBmBw7:-KhMwHRVAaoQNAFr1AI6:bjbud1y) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/README.md#L45)):*
> ```Markdown
> 
>  * Basic cars driving around on an open plane:
>    `bazel run drake/automotive:demo -- --simple_car_names=foo --num_trajectory_car=1`
> ```

Why is this a superior UX?

I agree with the need to turn off SimpleCar, but why does that mean making users who want it (which could be many) have to type in this `foo` silliness.

Possibly just a `num_simple_car` would be a better UX, still defaulted to 1?

---

*[drake/automotive/steering_command_driver.py, line 165 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMvF0xUi5XZntwT-Vz:-KhMvF0yMFcWt4lRg7dR:bbkx78b) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/steering_command_driver.py#L165)):*
> ```Python
>         pygame.init()
>         self.screen = pygame.display.set_mode((300, 70))
>         pygame.display.set_caption('Driving Command: ' + lcm_tag[16:])
> ```

`16` is magical.  Surely its derived from something?  I suspect `len("DRIVING_COMMAND_")`?

But this program doesn't (yet?) own the encapsulation of how that channel name is spelled, so it shouldn't be making assumptions about how long this value is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/automotive_demo.py, line 213 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMx59o5viKuGPL3Qfm:-KhNTUnAi6najDNa60n7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/automotive_demo.py#L213)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Missing docs, etc., etc.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.py, line 240 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMx9oQ1qwGbwhTlAWO:-KhNTmvdVEyqKmdhD5Jh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/automotive_demo.py#L240)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I think just `if args.simple_car_names:` would be more pythonic.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.py, line 245 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMwXLcsxRHI51wkS0Q:-KhNTo6OaFOKk8qOB38C:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/automotive_demo.py#L245)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI It would be more pythonic to rollup this condition with the line above with an `and`.
</blockquote></details>

Done.

---

*[drake/automotive/README.md, line 45 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMwHRUn5Yz9mUBmBw7:-KhNUD5DpzxFrGToq1OL:bd35v33) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/README.md#L45)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Why is this a superior UX?

I agree with the need to turn off SimpleCar, but why does that mean making users who want it (which could be many) have to type in this `foo` silliness.

Possibly just a `num_simple_car` would be a better UX, still defaulted to 1?
</blockquote></details>

Done. I defaulted `num_simple_car` to be zero to be consistent with all other `num_foo_car` parameters.

---

*[drake/automotive/steering_command_driver.py, line 165 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMvF0xUi5XZntwT-Vz:-KhNYzL9YcTOB-h8stgc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/steering_command_driver.py#L165)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`16` is magical.  Surely its derived from something?  I suspect `len("DRIVING_COMMAND_")`?

But this program doesn't (yet?) own the encapsulation of how that channel name is spelled, so it shouldn't be making assumptions about how long this value is.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-:-KhN_JoPcvo2cSI9tV5z:bblnx7v)*
<!-- Sent from Reviewable.io -->

rpoyner-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-:-KhNfCRh8Cw2-3SyKUXK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/automotive/automotive_demo.cc, line 99 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhNqCC6tHRCcs-SFdlV:-KhNqCC7lIVWwCURf00-:b9uqjnx) ([raw file](https://github.com/robotlocomotion/drake/blob/7fdc359b1e229de1b88a0a1fe2aa1ad4acd6cd95/drake/automotive/automotive_demo.cc#L99)):*
> ```cpp
>   const double kSimpleCarYSpacing{3};
> 
>   if (FLAGS_num_simple_car != 0 || !FLAGS_simple_car_names.empty()) {
> ```

The demo should fail-fast if a user sets _both_ num_simple_car and simple_car_names to be non-zero / non-empty.

---

*[drake/automotive/automotive_demo.cc, line 109 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhNqnLhkB0GBW9uhC-x:-KhNqnLiZEdWbKXTTb4Q:b-9zyz0n) ([raw file](https://github.com/robotlocomotion/drake/blob/7fdc359b1e229de1b88a0a1fe2aa1ad4acd6cd95/drake/automotive/automotive_demo.cc#L109)):*
> ```cpp
>       }
>     }
>     std::istringstream simple_car_name_stream(FLAGS_simple_car_names);
> ```

I think the argument here should be `simple_car_names`?

---

*[drake/automotive/automotive_demo.py, line 250 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhNrEbqWB-hkogY90wa:-KhNrEbrWBF2XtlcQzle:b-b7tnaq) ([raw file](https://github.com/robotlocomotion/drake/blob/7fdc359b1e229de1b88a0a1fe2aa1ad4acd6cd95/drake/automotive/automotive_demo.py#L250)):*
> ```Python
>                 args.simple_car_names += str(id)
>         if args.simple_car_names:
>             tail.append("--simple_car_names=" + args.simple_car_names)
> ```

This duplication of argument processing between here and the cc file is too brittle.

---

*[drake/automotive/automotive_demo.py, line 255 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhNr5QPTnoDU7bNgkWB:-KhNr5QQDkJX6_KyAhZw:bqt70v3) ([raw file](https://github.com/robotlocomotion/drake/blob/7fdc359b1e229de1b88a0a1fe2aa1ad4acd6cd95/drake/automotive/automotive_demo.py#L255)):*
> ```Python
>         if args.launch_steering_command_driver and args.simple_car_names != "":
>             name_list = args.simple_car_names.split(',')
>             for name in name_list:
> ```

I'm not sure that launching a keyboard UI for every user on a single computer is ever a useful conops?  Or at least, should not be the default one?

---

*[drake/automotive/steering_command_driver.py, line 165 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMvF0xUi5XZntwT-Vz:-KhNpvDOJqyucK-5_2Kz:b-p2ujib) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/steering_command_driver.py#L165)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Surely cutting at `_` is even more obscure and brittle?  Why shouldn't the UI be clear about what channel its transmitting on (i.e., show the entire string)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 99 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhNqCC6tHRCcs-SFdlV:-KhNxEAVRzgDkS04YSyw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7fdc359b1e229de1b88a0a1fe2aa1ad4acd6cd95/drake/automotive/automotive_demo.cc#L99)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The demo should fail-fast if a user sets _both_ num_simple_car and simple_car_names to be non-zero / non-empty.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 109 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhNqnLhkB0GBW9uhC-x:-KhNxJSkSjzzee6dqSnQ:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/7fdc359b1e229de1b88a0a1fe2aa1ad4acd6cd95/drake/automotive/automotive_demo.cc#L109)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think the argument here should be `simple_car_names`?
</blockquote></details>

Done. Thanks.

---

*[drake/automotive/automotive_demo.py, line 250 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhNrEbqWB-hkogY90wa:-KhO01QIEQ4ktamiipNA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7fdc359b1e229de1b88a0a1fe2aa1ad4acd6cd95/drake/automotive/automotive_demo.py#L250)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This duplication of argument processing between here and the cc file is too brittle.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.py, line 255 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhNr5QPTnoDU7bNgkWB:-KhNvqre2exIozLqMSGh:bmguwzv) ([raw file](https://github.com/robotlocomotion/drake/blob/7fdc359b1e229de1b88a0a1fe2aa1ad4acd6cd95/drake/automotive/automotive_demo.py#L255)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I'm not sure that launching a keyboard UI for every user on a single computer is ever a useful conops?  Or at least, should not be the default one?
</blockquote></details>

Done. I modified the program to support a dedicated command line flag called `---driving_command_gui_names` that allows the user to decide for which vehicles to launch the driving command GUI for.

---

*[drake/automotive/steering_command_driver.py, line 165 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhMvF0xUi5XZntwT-Vz:-KhNvoUSYP4wakP2NPSt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05f6045f72b8d52b2f2e57b3523645ba4e68c380/drake/automotive/steering_command_driver.py#L165)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Surely cutting at `_` is even more obscure and brittle?  Why shouldn't the UI be clear about what channel its transmitting on (i.e., show the entire string)?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 112 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhO1PjSMNTrys6aBPuX:-KhO1PjSMNTrys6aBPuY:b1kpdk5) ([raw file](https://github.com/robotlocomotion/drake/blob/c73168d235da4c4f62abd6f999af9e0fb000db6a/drake/automotive/automotive_demo.cc#L112)):*
> ```cpp
>       }
>     }
>     std::cout << "*** simple_car_names = " << simple_car_names << std::endl;
> ```

`cout`?

---

*[drake/automotive/automotive_demo.py, line 216 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhO5O3AjNIvkFejO2qB:-KhO5O3AjNIvkFejO2qC:b-t962xy) ([raw file](https://github.com/robotlocomotion/drake/blob/c73168d235da4c4f62abd6f999af9e0fb000db6a/drake/automotive/automotive_demo.py#L216)):*
> ```Python
>                         help="a comma separated list of vehicle names for " +
>                              "the vehicles for which a driving command GUI " +
>                              "should be launched")
> ```

This seems redundant with `launch_steering_command_driver`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774)*
<!-- Sent from Reviewable.io -->

rpoyner-tri commented:
Reviewed 1 of 1 files at r4, 1 of 4 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 112 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhO1PjSMNTrys6aBPuX:-KhOEJiQA35vtA5Gkb7a:bvu6wiw) ([raw file](https://github.com/robotlocomotion/drake/blob/c73168d235da4c4f62abd6f999af9e0fb000db6a/drake/automotive/automotive_demo.cc#L112)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`cout`?
</blockquote></details>

Oops. Removed.

---

*[drake/automotive/automotive_demo.py, line 216 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-KhO5O3AjNIvkFejO2qB:-KhOEPlwTD_o1uAyMrBO:b-3aou9w) ([raw file](https://github.com/robotlocomotion/drake/blob/c73168d235da4c4f62abd6f999af9e0fb000db6a/drake/automotive/automotive_demo.py#L216)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This seems redundant with `launch_steering_command_driver`?
</blockquote></details>

Done. I removed `launch_steering_command_driver`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> platform

I disclaim all responsibility for explaining to users why they have to say `--driving_command_gui_names=0` to run the simplest possible demo.

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-:-KhOoatCIzAv5sX0K-T7:b-wloyss)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I agree it's awkward and will fix that once I can think of a better UX strategy. Maybe we should remove the enumerated `SimpleCar` model names and require the user to spell them out.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5774#-:-KhOphqnNAOCbMGykXwd:b-w4xizr)*
<!-- Sent from Reviewable.io -->
